crates/agent/src/api/authorize_user_prefix.rs

```diff
@@ -0,0 +1,520 @@
+            ..Default::default()
+        })),
+        Err(Err(err)) => Err(err),
+    }
+}
+
+fn evaluate_authorization(
+    snapshot: &Snapshot,
+    user_id: uuid::Uuid,
+    user_email: Option<&String>,
+    prefix: &models::Prefix,
+    data_plane_name: &models::Name,
+    capability: models::Capability,
+) -> Result<
+    (
+        Option<chrono::DateTime<chrono::Utc>>,
+        (
+            jsonwebtoken::EncodingKey,
+            super::DataClaims, // Broker claims.
+            String,            // Broker address.
+            super::DataClaims, // Reactor claims.
+            String,            // Reactor address.
+        ),
+    ),
+    crate::api::ApiError,
+> {
+    if !tables::UserGrant::is_authorized(
+        &snapshot.role_grants,
+        &snapshot.user_grants,
+        user_id,
+        prefix,
+        capability,
+    ) {
+        return Err(anyhow::anyhow!(
+            "{} is not authorized to {prefix} for {capability:?}",
+            user_email.map(String::as_str).unwrap_or("user")
+        )
+        .with_status(StatusCode::FORBIDDEN));
+    }
+
+    if !tables::UserGrant::is_authorized(
+        &snapshot.role_grants,
+        &snapshot.user_grants,
+        user_id,
+        &data_plane_name,
+        models::Capability::Read,
+    ) {
+        return Err(anyhow::anyhow!(
+            "{} is not authorized to {data_plane_name}",
+            user_email.map(String::as_str).unwrap_or("user")
+        )
+        .with_status(StatusCode::FORBIDDEN));
+    }
+
+    let Some(data_plane) = snapshot.data_plane_by_catalog_name(&data_plane_name) else {
+        return Err(anyhow::anyhow!("data-plane {data_plane_name} not found")
+            .with_status(StatusCode::NOT_FOUND));
+    };
+    let Some(encoding_key) = data_plane.hmac_keys.first() else {
+        return Err(
+            anyhow::anyhow!("data-plane {data_plane_name} has no configured HMAC keys")
+                .with_status(StatusCode::INTERNAL_SERVER_ERROR),
+        );
+    };
+    let encoding_key = jsonwebtoken::EncodingKey::from_base64_secret(&encoding_key)
+        .context("invalid data-plane hmac key")?;
+
+    let broker_claims = super::DataClaims {
+        inner: proto_gazette::Claims {
+            cap: super::map_capability_to_gazette(capability),
+            exp: 0, // Filled later.
+            iat: 0, // Filled later.
+            iss: data_plane.data_plane_fqdn.clone(),
+            sub: user_id.to_string(),
+            sel: proto_gazette::broker::LabelSelector {
+                include: Some(labels::build_set([
+                    ("name:prefix", prefix.as_str()),
+                    ("name:prefix", &format!("recovery/capture/{prefix}")),
+                    ("name:prefix", &format!("recovery/derivation/{prefix}")),
+                    ("name:prefix", &format!("recovery/materialize/{prefix}")),
+                ])),
+                exclude: None,
+            },
+        },
+        prefixes: Vec::new(), // TODO(johnny): remove.
```

Comment:
nit: remove?

Reply:
can't until cronot / data-plane-gateway is gone.